### PR TITLE
feature: switch Azure AD graph API to Microsoft Graph API

### PR DESCRIPTION
**Pull request by johejo**
## Description

Microsoft strongly recommend switching `Azure AD Graph API` to `Microsoft Graph API`.
https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-graph-api

https://docs.microsoft.com/en-us/graph/api/user-get?view=graph-rest-1.0&tabs=http

Response of `GET https://graph.microsoft.com/v1.0/me` has field `mail` that same as the `Azure AD Graph API`.
So migration was easy.

```
HTTP/1.1 200 OK
Content-type: application/json
Content-length: 491

{
  "businessPhones": [
       "businessPhones-value"
   ],
   "displayName": "displayName-value",
   "givenName": "givenName-value",
   "jobTitle": "jobTitle-value",
   "mail": "mail-value",
   "mobilePhone": "mobilePhone-value",
   "officeLocation": "officeLocation-value",
   "preferredLanguage": "preferredLanguage-value",
   "surname": "surname-value",
   "userPrincipalName": "userPrincipalName-value",
   "id": "id-value"
}
```

But this change is **BRAKING CHANGE**.
Someone might use the token of Azure AD Graph API with `pass-access-token` option.
I will edit CHANGELOG.

## Motivation and Context

#354 

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,9 +6,11 @@
 - [#335] The session expiry for the OIDC provider is now taken from the Token Response (expires_in) rather than from the id_token (exp)
 
 ## Breaking Changes
+- [#440](https://github.com/pusher/oauth2_proxy/pull/440) Switch Azure AD Graph API to Microsoft Graph API (@johejo)
```
Comment (JoelSpeed) [inline CHANGELOG.md]: Would be good to expand on this and add a comment as to why this is a breaking change, similar to as we have done on previous releases

**Comment by johejo** [comment]
ProtectedResource will be change.
So, If application behind oauth2_proxy uses "X-Forwarded-Access-Token" for some Azure AD Graph API calling, after this change, those API calling will fail.
Behind applications have to change Azure AD Graph API to Microsoft Graph API if they use the token.

**Comment by JoelSpeed** [comment]
@johejo Ok that makes sense! Please expand on the breaking changelog entry and note that the existing behaviour can be retained if the `-resource` flag is set to the old value

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,9 +6,15 @@
...
+  Azure AD Graph API is deprecated. See [microsoft docs](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-graph-api).
+  Big Affected if application behind oauth2_proxy uses access token.
+  The behind application is recommended migrating to Microsoft Graph API.
+  Who want to continue using the Azure AD Graph API can use the `-resource` flag to keep the previous behavior.
```
Comment (JoelSpeed) [inline CHANGELOG.md]: I've reworded this a bit to help the flow, WDYT? Does this still capture you intention?
```suggestion
    - The Azure AD Graph API has been [deprecated](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-graph-api) and is being replaced by the Microsoft Graph API.
      If your application relies on the access token being passed to it to access the Azure AD Graph API,
  you should migrate you application to use the Microsoft Graph API.
      Existing behaviour can be retained by setting  `-resource=https://graph.windows.net`.
```

Comment (johejo) [reply CHANGELOG.md]: Good, thanks to review my poor English.

**Comment by johejo** [comment]
Sorry, approval was dismissed by my merge commit.
Ready for merge?

**Comment by syscll** [comment]
Would you mind fixing the conflict in the changelog? Thanks!

**Comment by johejo** [comment]
I fixed conflicts.

**Comment by JoelSpeed** [comment]
Are we considering this a breaking change? 🤔

**Comment by johejo** [comment]
> Are we considering this a breaking change? 🤔

Yes, This is my thought. https://github.com/oauth2-proxy/oauth2-proxy/pull/440#issuecomment-599022352
If my explanation is insufficient, should I modify the documentation a little more?

@JoelSpeed 
Do you have any questions? 
Do you think this is not a breaking change?

**Comment by JoelSpeed** [comment]
No sorry, I just needed to refresh myself. Agreed that this is breaking, fine by me to merge now since we already have other breaking changes going in in this next release anyway

**Pull request merged**